sdk/python/offlinehub.py

```diff
@@ -0,0 +1,163 @@
+        self._dims = dims
+        self._mapper = HnswlibIndexMapper()
+        self._size = 0
+        self._cap = 1024
+        self._idx.init_index(self._cap)
+        self.multiset(key_emb_iter)
+
+    def set(self, key, embedding):
+        """Set key to embedding.
+
+        Args:
+            key: The embedding index to set.
+            embedding: A python list representing the embedding vector to be
+            stored.
+        """
+        self._data[key] = embedding
+        idx = self._mapper.to_idx(key)
+        self._idx.add_items([embedding], [idx])
+        self._add_capacity_to_fit(1)
+
+    def get(self, key):
+        """Retrieves an embedding by key.
+    
+        Args:
+            key: The embedding index for retrieval.
+
+        Returns:
+            An embedding, which is a python list of floats.
+        """
+        return self._data[key]
+
+    def multiset(self, embedding_tuples):
+        """Set multiple embeddings at once.
+
+        Take a dictionary of key embedding pairs and set them in the index.
+
+        Args:
+            embedding_tuples: A dictionary from key to embedding or an iterator
+            of key, embedding pairs where key is a string and embedding is a
+            python list.
+        """
+        embeddings = []
+        idxs = []
+        if isinstance(embedding_tuples, Mapping):
+            embedding_tuples = embedding_tuples.items()
+
+        for key, embedding in embedding_tuples:
+            embeddings.append(embedding)
+            idxs.append(self._mapper.to_idx(key))
+            self._data[key] = embedding
+        # add_items fails if you give it two empty arrays.
+        if len(idxs) == 0:
+            return
+        self._add_capacity_to_fit(len(idxs))
+        self._idx.add_items(embeddings, idxs)
+
+    def multiget(self, keys):
+        """Get multiple embeddings at once.
+
+        Args:
+            keys: An iterator of embedding indices for retrieval.
+
+        Returns:
+            A list of embeddings.
+        """
+        return [self._data[key] for key in keys]
+
+    def nearest_neighbor(self, key, num):
+        """Finds N nearest neighbors for a given embedding record.
+
+        Args:
+            key: The embedding index to get the nearest neighbors of.
+            num: The number of nearest neighbors.
+
+        Returns:
+            A num size list of embedding vectors that are closest to the
+            provided embedding.
+        """
+        idx = self._mapper.to_idx(key)
+        # knn_query expects a list of indices to work in batch.
+        idxs = [idx]
+        # knn_query returns itself, so we get one extra nearest neighbor and drop
+        # idx when mapping the indices.
+        list_results, list_distances = self._idx.knn_query(idxs, num + 1)
+        results = list_results[0]
+        return [
+            self._mapper.to_key(result) for result in results if result != idx
+        ]
+
+    def _add_capacity_to_fit(self, size):
```

Comment:
Is self._size changed?

Reply:
Great catch, I'll write a test for this as well